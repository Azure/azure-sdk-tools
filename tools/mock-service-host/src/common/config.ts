export interface Config {
    httpsPortStateful: number
    httpPortStateless: number
    httpsPortStateless: number
    internalErrorPort: number
    loggingConsoleLevel: string
    serviceEnvironment: string
    serviceName: string
    specRetrievalMethod: string
    specRetrievalGitBranch: string
    specRetrievalGitCommitID: string
    specRetrievalGitUrl: string
    specRetrievalLocalRelativePath: string
    specRetrievalRefreshEnabled: boolean
    specRetrievalGitAuthMode: string
    specRetrievalGitAuthToken: string
    specRetrievalRefreshIntervalMilliseconds: number
    validationPathsPattern: string[]
    enableExampleGeneration: boolean
    exampleGenerationFolder: string
    cascadeEnabled: boolean
}
