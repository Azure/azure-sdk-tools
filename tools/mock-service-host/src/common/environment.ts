export enum Environment {
    Development = 'development',
    Production = 'production',
    Test = 'test'
}

export enum SpecRetrievalAuthMode {
    None = 'none',
    Token = 'token'
}

export enum SpecRetrievalMethod {
    Filesystem = 'filesystem',
    Git = 'git'
}
