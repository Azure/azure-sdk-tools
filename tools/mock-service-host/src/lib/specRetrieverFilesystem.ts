import { BaseSpecRetriever, SpecRetriever } from './specRetriever'
import { Config } from '../common/config'
import { InjectableTypes } from './injectableTypes'
import { inject, injectable } from 'inversify'
import { logger } from '../common/utils'

/**
 * Copies specifications from a specified location on the filesystem and caches them locally.
 */
@injectable()
export class SpecRetrieverFilesystem extends BaseSpecRetriever implements SpecRetriever {
    constructor(@inject(InjectableTypes.Config) config: Config) {
        super(config)
    }

    protected async retrieveSpecsImpl(): Promise<void> {
        logger.info(
            `SpecRetrieverFilesystem: Set spec cache path as ${this.config.specRetrievalLocalRelativePath}`
        )
        this.localPath = this.config.specRetrievalLocalRelativePath
    }
}
