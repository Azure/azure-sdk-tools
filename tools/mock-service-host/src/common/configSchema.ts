import { Config } from './config'
import { Environment, SpecRetrievalAuthMode, SpecRetrievalMethod } from './environment'

import convict from 'convict'
/**
 * Service configuration schema.
 */
export const configSchema = convict<Config>({
    httpsPortStateful: {
        doc: 'The http port to use for stateful status.',
        format: 'port',
        default: 8441,
        env: 'httpsPortStateful'
    },
    httpPortStateless: {
        doc: 'The http port to use for stateless status.',
        format: 'port',
        default: 8442,
        env: 'httpPortStateless'
    },
    httpsPortStateless: {
        doc: 'The https port to use for stateful status.',
        format: 'port',
        default: 8443,
        env: 'httpsPortStateless'
    },
    internalErrorPort: {
        doc: 'The https port to use for internal errors.',
        format: 'port',
        default: 8445,
        env: 'internalErrorPort'
    },
    loggingConsoleLevel: {
        doc: 'The logging level to use for console output.',
        format: ['error', 'warn', 'info', 'verbose', 'debug', 'silly'],
        default: 'info',
        env: 'loggingConsoleLevel'
    },
    serviceEnvironment: {
        doc: 'The application environment.',
        format: [Environment.Production, Environment.Test, Environment.Development],
        default: Environment.Development,
        env: 'NODE_ENV'
    },
    serviceName: {
        doc: 'The name of the service.',
        format: String,
        default: 'validate-v2',
        env: 'serviceName'
    },
    specRetrievalMethod: {
        doc: 'The method to use for retrieving specifications',
        format: [SpecRetrievalMethod.Git, SpecRetrievalMethod.Filesystem],
        default: SpecRetrievalMethod.Git,
        env: 'specRetrievalMethod'
    },
    specRetrievalGitBranch: {
        doc: 'The branch to use when pulling specs from a git repo.',
        format: String,
        default: 'main',
        env: 'specRetrievalGitBranch'
    },
    specRetrievalGitCommitID: {
        doc: 'The commitid to use when pulling specs from a git repo.',
        format: String,
        default: '',
        env: 'specRetrievalGitCommitID'
    },
    specRetrievalGitUrl: {
        doc:
            'The repo URL to use when pulling specs from git. To authenticate with GitHub, use https and provide values for authorization header (user name, personal access token) in the URL, e.g. https://<user>:<token>@<repo>',
        format: String,
        default: 'https://github.com/Azure/azure-rest-api-specs',
        env: 'specRetrievalGitUrl'
    },
    specRetrievalLocalRelativePath: {
        doc:
            "The relative path for caching specifications underneath the service's current working directory. NOTE: This directory will be deleted if it exists.",
        format: String,
        default: 'specificationCache',
        env: 'specRetrievalLocalRelativePath'
    },
    specRetrievalRefreshEnabled: {
        doc:
            'If true the validator will refresh specifications on an interval specified by specRetrievalRefreshIntervalMilliseconds.',
        format: Boolean,
        default: true,
        env: 'specRetrievalRefreshEnabled'
    },
    specRetrievalRefreshIntervalMilliseconds: {
        doc:
            'How often to refresh the validator from the source sepcifications, specified in milliseconds.',
        format: Number,
        default: 3600000,
        env: 'specRetrievalRefreshIntervalMilliseconds'
    },
    specRetrievalGitAuthMode: {
        doc: 'The authentication mode to use when pulling from a GitHub repo',
        format: [SpecRetrievalAuthMode.None, SpecRetrievalAuthMode.Token],
        default: SpecRetrievalAuthMode.None,
        env: 'specRetrievalGitAuthMode'
    },
    specRetrievalGitAuthToken: {
        doc: 'The personal access token to use when using GitHub token authentication',
        format: String,
        default: '',
        env: 'specRetrievalGitAuthToken',
        sensitive: true
    },
    validationPathsPattern: {
        doc: 'The pattern which identifies cached specs to use in validation.',
        format: Array,
        default: ['/specification/**/resource-manager/**/*.json'],
        env: 'validationPathsPattern'
    },
    excludedValidationPathsPattern: {
        doc: 'The pattern which excluded from the validationPathsPattern',
        format: Array,
        default: [
            '**/examples/**/*',
            '**/quickstart-templates/**/*',
            '**/schema/**/*',
            '**/live/**/*',
            '**/wire-format/**/*',
            '**/resource-manager/*/*/*/scenarios/**'
        ],
        env: 'excludedValidationPathsPattern'
    },
    enableExampleGeneration: {
        doc: 'If true example files will not be generated along with each REST calling.',
        format: Boolean,
        default: false,
        env: 'enableExampleGeneration'
    },
    exampleGenerationFolder: {
        doc: 'The folder name used for example generation for each REST calling.',
        format: String,
        default: 'mock',
        env: 'exampleGenerationFolder',
        sensitive: true
    },
    cascadeEnabled: {
        doc:
            'If true resources in different levels should be created orderly and can be deleted cascadingly.',
        format: Boolean,
        default: false,
        env: 'cascadeEnabled'
    }
})
