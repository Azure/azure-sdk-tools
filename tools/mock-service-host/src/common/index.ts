import 'dotenv/config'

import { Config } from './config'
import { Environment } from './environment'
import { configSchema } from './configSchema'
import { environmentConfigDevelopment } from './environmentConfigDevelopment'
import { environmentConfigProduction } from './environmentConfigProduction'
import { environmentConfigTest } from './environmentConfigTest'

const env = configSchema.get('serviceEnvironment') as Environment

const environmentOverrides: any = {
    [Environment.Production]: environmentConfigProduction,
    [Environment.Test]: environmentConfigTest,
    [Environment.Development]: environmentConfigDevelopment
}

// Load environment dependent configuration
configSchema.load(environmentOverrides[env])

// Perform validation
configSchema.validate({ allowed: 'strict' })

export const config: Config = configSchema.getProperties()
