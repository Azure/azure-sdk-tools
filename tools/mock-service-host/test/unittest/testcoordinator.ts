import 'reflect-metadata' // Must be imported exactly once for Inversify

/* eslint-disable @typescript-eslint/no-var-requires */
import * as assert from 'assert'
import * as path from 'path'
import { Coordinator } from '../../src/mid/coordinator'
import {
    ExampleNotFound,
    ExampleNotMatch,
    HttpStatusCode,
    IntentionalError,
    LroCallbackNotFound,
    NoParentResource,
    ResourceNotFound,
    ValidationFail
} from '../../src/common/errors'
import { RequestResponsePair } from 'oav/dist/lib/liveValidation/liveValidator'
import { ResponseGenerator } from '../../src/mid/responser'
import { SpecRetrieverFilesystem } from '../../src/lib/specRetrieverFilesystem'
import { config } from '../../src/common/index'
import {
    createLiveRequestForCreateApiManagementService,
    createLiveRequestForCreateRG,
    createLiveRequestForDeleteApiManagementService,
    mockDefaultResponse,
    mockRequest
} from '../tools'

const statefulProfile = {
    stateful: true
}
const statelessProfile = {
    stateful: false
}
const alwaysErrorProfile = {
    alwaysError: 500
}

describe('initialize validator:', () => {
    const specRetriever = new SpecRetrieverFilesystem(config)
    const coordinator = new Coordinator(config, specRetriever, new ResponseGenerator(config))

    beforeAll(async () => {
        await coordinator.initialize()
    })

    it('should initialize with swagger directory', async () => {
        const cache = coordinator.Validator.operationSearcher.cache
        assert.strictEqual(true, cache.has('microsoft.apimanagement'))
        assert.strictEqual(true, cache.has('microsoft.media'))
    })
})

describe('generateResponse()', () => {
    const specRetriever = new SpecRetrieverFilesystem(config)
    const responseGenerator = new ResponseGenerator(config)
    const coordinator = new Coordinator(config, specRetriever, responseGenerator)
    beforeAll(async () => {
        await coordinator.initialize()
    })

    beforeEach(async () => {
        responseGenerator.initiateResourcePool()
        jest.restoreAllMocks()
    })

    it('validate GET input', async () => {
        const fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input.json')
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        await coordinator.generateResponse(request, response, statelessProfile)
        assert.strictEqual(response.statusCode, pair.liveResponse.statusCode)
        pair.liveResponse.body = response.body
        pair.liveResponse.headers = response.headers

        await coordinator.Validator.validateLiveRequestResponse(pair)
    })

    it('alwaysError: return 500 even for a valid_input', async () => {
        const fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input.json')
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        await expect(
            coordinator.generateResponse(request, response, alwaysErrorProfile)
        ).rejects.toThrow(IntentionalError)
    })

    it('validate DELETE input', async () => {
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_delete.json'
        )
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        await coordinator.generateResponse(request, response, statelessProfile)
        assert.strictEqual(response.statusCode, pair.liveResponse.statusCode)
        expect(response).toMatchSnapshot()
        pair.liveResponse.body = response.body
        pair.liveResponse.headers = response.headers
        pair.liveResponse.headers['Content-Type'] = 'application/json'
        const result = await coordinator.Validator.validateLiveRequestResponse(pair)
        assert.ok(result.requestValidationResult.isSuccessful)
        assert.ok(result.responseValidationResult.isSuccessful)
    })

    it('invalidate input', async () => {
        const fileName = path.join(__dirname, '..', 'testData', 'payloads', 'invalidUrl_input.json')
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        await expect(
            coordinator.generateResponse(request, response, statelessProfile)
        ).rejects.toThrow(ValidationFail)
    })

    it('special rule of GET resourceGroup', async () => {
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'special_resourcegroup.json'
        )
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        await coordinator.generateResponse(request, response, statelessProfile)
        assert.strictEqual(response.statusCode, pair.liveResponse.statusCode)
        expect(response).toMatchSnapshot()
    })

    it('special rule of GET locations', async () => {
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'special_resourcegroup.json'
        )
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        await coordinator.generateResponse(request, response, statelessProfile)
        assert.strictEqual(response.statusCode, pair.liveResponse.statusCode)
        expect(response).toMatchSnapshot()
        if (!pair.liveResponse.headers) pair.liveResponse.headers = {}
        const result = await coordinator.Validator.validateLiveRequestResponse(pair)
        assert.strictEqual(result.requestValidationResult.isSuccessful, undefined) // since this is a special URI not handled by oav
    })

    it('findLROGet', async () => {
        jest.spyOn(coordinator.liveValidator, 'parseValidationRequest').mockImplementation(
            (
                requestUrl: string,
                requestMethod: string | undefined | null,
                correlationId: string
            ) => {
                const getUrls = [
                    'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type',
                    'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType',
                    'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2',
                    'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2',
                    'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2/start',
                    'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2/Type3'
                ]
                if (getUrls.indexOf(requestUrl.split('?')[0]) < 0) {
                    throw new Error('No operation')
                }
                return {
                    providerNamespace: '',
                    resourceType: '',
                    apiVersion: '',
                    requestMethod: 'get',
                    host: '',
                    pathStr: '',
                    correlationId: correlationId,
                    requestUrl: requestUrl
                }
            }
        )
        jest.spyOn(coordinator.liveValidator.operationSearcher, 'search').mockImplementation(
            (_) => {
                return {
                    operationMatch: 'fake',
                    apiVersion: 'fake'
                } as any
            }
        )

        const liveRequest = {
            protocol: 'https',
            url:
                '/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2/Type3/myType3/start?api-version=20210701',
            method: 'POST',
            headers: {
                host: 'localhost:8443'
            },
            localPort: 8443
        }
        assert.strictEqual(
            await coordinator.findLROGet(liveRequest),
            'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2?api-version=20210701&lro-callback=true'
        )

        liveRequest.url =
            '/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2?api-version=20210701'
        assert.strictEqual(
            await coordinator.findLROGet(liveRequest),
            'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2?api-version=20210701&lro-callback=true'
        )

        liveRequest.url =
            '/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2/stop?api-version=20210701'
        assert.strictEqual(
            await coordinator.findLROGet(liveRequest),
            'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/Type2/myType2?api-version=20210701&lro-callback=true'
        )

        liveRequest.url =
            '/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType/new?api-version=20210701'
        assert.strictEqual(
            await coordinator.findLROGet(liveRequest),
            'https://localhost:8443/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type/myType?api-version=20210701&lro-callback=true'
        )

        liveRequest.url =
            '/subscriptions/xxx/resourceGroups/xx/providers/Microsoft.Mock/Type2/myType/new?api-version=20210701'
        expect(coordinator.findLROGet(liveRequest)).rejects.toThrow(LroCallbackNotFound)
    })

    it("degrade to non-lro if can't find callback url", async () => {
        const response = mockDefaultResponse()
        const fileName = path.join(__dirname, '..', 'testData', 'payloads', 'restore_service.json')
        const pair: RequestResponsePair = require(fileName)
        if (!pair.liveRequest.headers) {
            pair.liveRequest.headers = {}
        }
        const request = mockRequest(pair.liveRequest)
        await coordinator.generateResponse(request, response, statelessProfile)
        expect(response).toMatchSnapshot()
    })

    it("return error if can't find callback url and there is no 200 response", async () => {
        const response = mockDefaultResponse()
        const fileName = path.join(__dirname, '..', 'testData', 'payloads', 'backup_service.json')
        const pair: RequestResponsePair = require(fileName)
        if (!pair.liveRequest.headers) {
            pair.liveRequest.headers = {}
        }
        const request = mockRequest(pair.liveRequest)
        await expect(
            coordinator.generateResponse(request, response, statelessProfile)
        ).rejects.toThrow(LroCallbackNotFound)
    })

    it('end up with example response', async () => {
        const response = mockDefaultResponse()
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'create_api_release.json'
        )
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        await coordinator.generateResponse(request, response, statelessProfile)
        expect(response).toMatchSnapshot()
    })

    it('end up with mocked response', async () => {
        const response = mockDefaultResponse()
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'create_api_operation_policy.json'
        )
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        await coordinator.generateResponse(request, response, statelessProfile)
        assert.strictEqual(response.statusCode, '200')
        assert.notStrictEqual(
            (response.body as any)['properties']['policyContent'],
            '<policies></policies>'
        )
    })
})

describe('genStatefulResponse()', () => {
    const specRetriever = new SpecRetrieverFilesystem(config)
    const responseGenerator = new ResponseGenerator(config)
    const coordinator = new Coordinator(config, specRetriever, responseGenerator)

    beforeAll(async () => {
        await coordinator.initialize()
        config.cascadeEnabled = true
    })

    beforeEach(async () => {
        responseGenerator.initiateResourcePool()
    })

    it('stateful: return 404 for GET even it is a valid_input', async () => {
        const fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input.json')
        const pair: RequestResponsePair = require(fileName)

        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        //await generateResponse(validator, request, response, statelessProfile)

        expect(
            responseGenerator.genStatefulResponse(
                request,
                response,
                statefulProfile,
                response.statusCode,
                response.body
            )
        ).rejects.toThrow(ResourceNotFound)
    })

    it('stateful: return 404 for DELETE even it is a valid_input', async () => {
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_delete.json'
        )
        const pair: RequestResponsePair = require(fileName)

        const request = mockRequest(pair.liveRequest)
        const response = mockDefaultResponse()
        //await generateResponse(validator, request, response, statelessProfile)

        expect(
            responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
                body: 'faked 200'
            })
        ).rejects.toThrow(ResourceNotFound)
    })

    it('stateful create->read->delete', async () => {
        const response = mockDefaultResponse()

        //create rg and service before test
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateRG()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateApiManagementService()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )

        // create
        let fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input_create.json')
        let pair: RequestResponsePair = require(fileName)
        let request = mockRequest(pair.liveRequest)
        await responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
            body: 'faked 200'
        })
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())

        // read
        fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input.json')
        pair = require(fileName)
        request = mockRequest(pair.liveRequest)
        await responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
            body: 'faked 200'
        })
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())

        // delete
        fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input_delete.json')
        pair = require(fileName)
        request = mockRequest(pair.liveRequest)
        await responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
            body: 'faked 200'
        })
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())
    })

    it('stateful with cascadeEnabled: create subresource should be failed if parent resource has not been created', async () => {
        const response = mockDefaultResponse()

        //create rg
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateRG()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )

        // create resource user without create it's parent resource service
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_create.json'
        )
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        expect(
            responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
                body: 'faked 200'
            })
        ).rejects.toThrow(NoParentResource)
    })

    it('stateful with cascadeEnabled: deleting a parent resource will also delete its child', async () => {
        const response = mockDefaultResponse()

        //create rg and service before test
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateRG()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateApiManagementService()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )

        // create
        let fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input_create.json')
        let pair: RequestResponsePair = require(fileName)
        let request = mockRequest(pair.liveRequest)
        await responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
            body: 'faked 200'
        })
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())

        // delete parent resource
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForDeleteApiManagementService()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())

        // reading the child should be failed
        fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input.json')
        pair = require(fileName)
        request = mockRequest(pair.liveRequest)
        expect(
            responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
                body: 'faked 200'
            })
        ).rejects.toThrow(ResourceNotFound)
    })
})

describe('genStatefulResponse() with cascadeEnabled==false', () => {
    const specRetriever = new SpecRetrieverFilesystem(config)
    const responseGenerator = new ResponseGenerator(config)
    const coordinator = new Coordinator(config, specRetriever, responseGenerator)

    beforeAll(async () => {
        config.cascadeEnabled = false
        await coordinator.initialize()
    })

    beforeEach(async () => {
        responseGenerator.initiateResourcePool()
    })

    it('stateful with cascadeEnabled==false: create subresource can succeed even if parent resource has not been created', async () => {
        const response = mockDefaultResponse()

        //create rg
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateRG()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )

        // create resource user without create it's parent resource service
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_create.json'
        )
        const pair: RequestResponsePair = require(fileName)
        const request = mockRequest(pair.liveRequest)
        await responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
            body: 'faked 200'
        })
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())
    })

    it('stateful with cascadeEnabled==false: deleting a parent resource will not delete its child', async () => {
        const response = mockDefaultResponse()

        //create rg and service before test
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateRG()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForCreateApiManagementService()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )

        // create
        let fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input_create.json')
        let pair: RequestResponsePair = require(fileName)
        let request = mockRequest(pair.liveRequest)
        await responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
            body: 'faked 200'
        })
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())

        // delete parent resource
        await responseGenerator.genStatefulResponse(
            mockRequest(createLiveRequestForDeleteApiManagementService()),
            response,
            statefulProfile,
            '200',
            {
                body: 'faked 200'
            }
        )
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())

        // reading the child should be failed
        fileName = path.join(__dirname, '..', 'testData', 'payloads', 'valid_input.json')
        pair = require(fileName)
        request = mockRequest(pair.liveRequest)
        await responseGenerator.genStatefulResponse(request, response, statefulProfile, '200', {
            body: 'faked 200'
        })
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())
    })
})

describe('Example Validation', () => {
    const specRetriever = new SpecRetrieverFilesystem(config)
    const responseGenerator = new ResponseGenerator(config)
    const coordinator = new Coordinator(config, specRetriever, responseGenerator)

    beforeAll(async () => {
        config.cascadeEnabled = false
        await coordinator.initialize()
    })

    beforeEach(async () => {
        responseGenerator.initiateResourcePool()
    })

    it('valid example request', async () => {
        const response = mockDefaultResponse()

        // create resource user without create it's parent resource service
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_create.json'
        )
        const pair: RequestResponsePair = require(fileName)
        if (!pair.liveRequest.headers) {
            pair.liveRequest.headers = {}
        }
        pair.liveRequest.headers['example-id'] = 'ApiManagementCreateUserBasic'
        const request = mockRequest(pair.liveRequest)
        await coordinator.generateResponse(request, response, statelessProfile)
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())
        expect(response.body?.name).toMatch('aaaaaaaaaaaaaaaaaaaaaaaaaaaa')

        // verify behaviour when don't set example-id
        request.headers = {}
        await coordinator.generateResponse(request, response, statelessProfile)
        assert.strictEqual(response.statusCode, HttpStatusCode.OK.toString())
        expect(response.body?.name).toMatch('randomUsers')
    })

    it('will fail if send wrong example-id', async () => {
        const response = mockDefaultResponse()

        // create resource user without create it's parent resource service
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_create.json'
        )
        const pair: RequestResponsePair = require(fileName)
        if (!pair.liveRequest.headers) {
            pair.liveRequest.headers = {}
        }
        pair.liveRequest.headers['example-id'] = 'Wrong-ApiManagementCreateUserBasic'
        const request = mockRequest(pair.liveRequest)
        await expect(
            coordinator.generateResponse(request, response, statelessProfile)
        ).rejects.toThrow(ExampleNotFound)
    })

    it("will fail if request body don't match example", async () => {
        const response = mockDefaultResponse()

        // create resource user without create it's parent resource service
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_create.json'
        )
        const pair: RequestResponsePair = require(fileName)
        if (!pair.liveRequest.headers) {
            pair.liveRequest.headers = {}
        }
        pair.liveRequest.headers['example-id'] = 'ApiManagementCreateUserBasic'
        if (!pair.liveRequest.body) {
            pair.liveRequest.body = {}
        }
        ;(pair.liveRequest.body as any)['properties']['firstName'] = 'fooName'
        const request = mockRequest(pair.liveRequest)
        await expect(
            coordinator.generateResponse(request, response, statelessProfile)
        ).rejects.toThrow(ExampleNotMatch)
    })

    it("will fail if request header don't match example", async () => {
        const response = mockDefaultResponse()

        // create resource user without create it's parent resource service
        const fileName = path.join(
            __dirname,
            '..',
            'testData',
            'payloads',
            'valid_input_create.json'
        )
        const pair: RequestResponsePair = require(fileName)
        if (!pair.liveRequest.headers) {
            pair.liveRequest.headers = {}
        }
        pair.liveRequest.headers['example-id'] = 'ApiManagementCreateUserBasic'
        pair.liveRequest.headers['If-Match'] = 'fakedValue'
        const request = mockRequest(pair.liveRequest)
        await expect(
            coordinator.generateResponse(request, response, statelessProfile)
        ).rejects.toThrow(ExampleNotMatch)
    })
})
