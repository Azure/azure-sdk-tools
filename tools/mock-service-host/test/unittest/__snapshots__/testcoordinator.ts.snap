// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateResponse() degrade to non-lro if can't find callback url 1`] = `
VirtualServerResponse {
  "body": Object {
    "etag": "AAAAAAAXX6Y=",
    "id": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.ApiManagement/service/apimService1",
    "location": "East Asia",
    "name": "restore",
    "properties": Object {
      "createdAtUtc": "2015-09-22T01:50:34.7921566Z",
      "gatewayUrl": "https://apimService1.azure-api.net",
      "hostnameConfigurations": Array [],
      "managementApiUrl": "https://apimService1.management.azure-api.net",
      "portalUrl": "https://apimService1.portal.azure-api.net",
      "provisioningState": "Succeeded",
      "publicIPAddresses": Array [
        "207.46.155.24",
      ],
      "publisherEmail": "admin@live.com",
      "publisherName": "Contoso",
      "scmUrl": "https://apimService1.scm.azure-api.net",
      "targetProvisioningState": "",
      "virtualNetworkType": "None",
    },
    "sku": Object {
      "capacity": 1,
      "name": "Premium",
    },
    "tags": Object {
      "UID": "52ed5986-717b-45b4-b17c-3df8db372cff",
    },
    "type": "Microsoft.ApiManagement/serviceFaked/restore",
  },
  "headers": Object {},
  "statusCode": "200",
}
`;

exports[`generateResponse() end up with example response 1`] = `
VirtualServerResponse {
  "body": Object {
    "id": "/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.ApiManagement/service/apimService1/apis/a1/releases/testrev",
    "name": "testrev",
    "properties": Object {
      "apiId": "/subscriptions/subid/resourceGroups/rg1/providers/Microsoft.ApiManagement/service/apimService1/apis/a1",
      "createdDateTime": "2018-02-08T20:52:00.65Z",
      "notes": "yahooagain",
      "updatedDateTime": "2018-02-08T20:52:00.65Z",
    },
    "type": "Microsoft.ApiManagement/service/apis/releases",
  },
  "headers": Object {},
  "statusCode": "200",
}
`;

exports[`generateResponse() special rule of GET locations 1`] = `
VirtualServerResponse {
  "body": Object {
    "id": "/subscriptions/randomSub/resourceGroups/randomRG",
    "location": "eastus",
    "managedBy": null,
    "name": "randomRG",
    "properties": Object {
      "provisioningState": "Succeeded",
    },
    "tags": Object {},
    "type": "Microsoft.Resources/resourceGroups",
  },
  "headers": Object {},
  "statusCode": "200",
}
`;

exports[`generateResponse() special rule of GET resourceGroup 1`] = `
VirtualServerResponse {
  "body": Object {
    "id": "/subscriptions/randomSub/resourceGroups/randomRG",
    "location": "eastus",
    "managedBy": null,
    "name": "randomRG",
    "properties": Object {
      "provisioningState": "Succeeded",
    },
    "tags": Object {},
    "type": "Microsoft.Resources/resourceGroups",
  },
  "headers": Object {},
  "statusCode": "200",
}
`;

exports[`generateResponse() validate DELETE input 1`] = `
VirtualServerResponse {
  "body": undefined,
  "headers": Object {},
  "statusCode": "200",
}
`;
